beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	return nil
+}
+
+// deleteCommitteesFrom deletes committees starting from the given period.
+func (s *CommitteeChain) deleteCommitteesFrom(batch ethdb.Batch, period uint64) {
+	deleted := s.committees.deleteFrom(batch, period)
+	for period := deleted.First; period < deleted.AfterLast; period++ {
+		s.syncCommitteeCache.Remove(period)
+	}
+}
+
+// GetCommittee returns the committee at the given period.
+// Note: GetCommittee can be called either with locked or unlocked chain mutex.
+func (s *CommitteeChain) GetCommittee(period uint64) *types.SerializedSyncCommittee {
+	committee, _ := s.committees.get(period)
+	return committee
+}
+
+// AddCommittee adds a committee at the given period if possible.
+func (s *CommitteeChain) AddCommittee(period uint64, committee *types.SerializedSyncCommittee) error {
+	s.chainmu.Lock()
+	defer s.chainmu.Unlock()
+
+	if !s.committees.periods.CanExpand(period) {
+		return ErrInvalidPeriod
+	}
+	root := s.getCommitteeRoot(period)
+	if root == (common.Hash{}) {
+		return ErrInvalidPeriod
+	}
+	if root != committee.Root() {
+		return ErrWrongCommitteeRoot
+	}
+	if !s.committees.periods.Includes(period) {
+		if err := s.committees.add(s.db, period, committee); err != nil {
+			return err
+		}
+		s.syncCommitteeCache.Remove(period)
+	}
+	return nil
+}
+
+// GetUpdate returns the update at the given period.
+// Note: GetUpdate can be called either with locked or unlocked chain mutex.
+func (s *CommitteeChain) GetUpdate(period uint64) *types.LightClientUpdate {
+	update, _ := s.updates.get(period)
+	return update
+}
+
+// InsertUpdate adds a new update if possible.
+func (s *CommitteeChain) InsertUpdate(update *types.LightClientUpdate, nextCommittee *types.SerializedSyncCommittee) error {
+	s.chainmu.Lock()
+	defer s.chainmu.Unlock()
+
+	period := update.AttestedHeader.Header.SyncPeriod()
+	if !s.updates.periods.CanExpand(period) || !s.committees.periods.Includes(period) {
+		return ErrInvalidPeriod
+	}
+	if s.minimumUpdateScore.BetterThan(update.Score()) {
+		return ErrInvalidUpdate
+	}
+	oldRoot := s.getCommitteeRoot(period + 1)
+	reorg := oldRoot != (common.Hash{}) && oldRoot != update.NextSyncCommitteeRoot
+	if oldUpdate, ok := s.updates.get(period); ok && !update.Score().BetterThan(oldUpdate.Score()) {
+		// a better or equal update already exists; no changes, only fail if new one tried to reorg
+		if reorg {
+			return ErrCannotReorg
+		}
+		return nil
+	}
+	if s.fixedRoots.periods.Includes(period+1) && reorg {
+		return ErrCannotReorg
+	}
+	if !s.verifyUpdate(update) {
+		return ErrInvalidUpdate
+	}
+	addCommittee := !s.committees.periods.Includes(period+1) || reorg
+	if addCommittee {
+		if nextCommittee == nil {
+			return ErrNeedCommittee
+		}
+		if nextCommittee.Root() != update.NextSyncCommitteeRoot {
+			return ErrWrongCommitteeRoot
+		}
+	}
+	batch := s.db.NewBatch()
+	if reorg {
```

Comment:
Reorg is not suitable in this context, replace the next committee specified by the update is more suitable.

IIUC, we can either extend the update chain with a new one, or replace the last one, but we can not allowed to replace a middle one right?

Reply:
We can replace an update in the middle with one that has a higher signer participation. In a more likely scenario this just means we add a better update that proves the same next committee root. Reorgs of the committee chain are also theoretically possible, though hopefully they will not happen on mainnet. One thing to consider is that wrong sync committee signatures are currently not slashable (hopefully they will be at one point, at least under certain conditions). So it is not entirely impossible to organize an attack and create a valid update that leads to a fake chain. Still, it's a lot less likely to create one with more signers than the actual real updates (which have over 90% participation), and the whole attack is a lot less likely to happen if clients can immediately recover and propagate the real updates.